lib/Runtime/Library/JSON.cpp

```diff
@@ -688,55 +689,45 @@ namespace JSON
 
                         result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
                         Js::DynamicTypeHandler * typeHandler = dynamicObject->GetTypeHandler();
-                        enumerator.Reset();
-                        Js::PropertyId previousId;
+
                         // if object has an objectarray, (aka indexed properties)
                         // we need to loop them first since they won't show up on the second loop below
                         if (dynamicObject->HasObjectArray())
                         {
-                            int totalNumPropertyCount = this->GetPropertyCount(object, &enumerator) - propertyCount;
-                            Assert(totalNumPropertyCount >= 0);
-
-                            int index = 0;
                             enumerator.Reset();
-                            while (index < totalNumPropertyCount && (propertyName = enumerator.MoveAndGetNext(id)) != NULL)
+                            while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL)
                             {
                                 // if unsuccessful get propertyId from the string
-                                scriptContext->GetOrAddPropertyRecord(propertyName->GetString(), propertyName->GetLength(), &propRecord);
+                                scriptContext->GetOrAddPropertyRecord(propertyName, &propRecord);
+                                if (!propRecord->IsNumeric()) break;
```

Comment:
Do we guarantee that numeric properties always come first in enumeration order? Is that enforced by spec, or an implementation detail?

Reply:
if we don't https://github.com/Microsoft/ChakraCore/pull/3894/files#diff-8092540edf441c597a868bfdb4789f0aR58 will break.